File: impl/java/src/main/java/sh/bubblefish/naalp/Carriage.java
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp;

import java.util.List;

/**
 * C12 — foreign carriage by class for the Java SDK (design.md §13; R-14.1..14.8, R-18.6).
 *
 * <p>N-AALP carries a foreign agent protocol by wrapping its message, octet-for-octet, in a signed
 * N-AALP carriage object whose effect, safety, identity, and audit apply, and whose foreign body is
 * interpreted by a carriage CLASS — not a bespoke per-protocol mapping (R-14.1). There are five
 * structured classes (JSONRPC, HTTP, MSG, STREAM, DOC) plus a universal OPAQUE class that makes any
 * protocol — including one nobody has defined — carriable immediately on an experimental protocol id
 * with no registration (R-14.1, R-18.6). The foreign field is carried VERBATIM and MUST NOT be
 * re-serialized, canonicalized, summarized, or rewritten (R-14.4); N-AALP metadata is carried around
 * it, never inside it. The carriage object's signer remains the authority — a foreign identity never
 * becomes an N-AALP authorization identity (R-14.6).
 *
 * <p>An independent transcription of impl/go/carriage (cross-checked against
 * impl/python/naalp/carriage). Each class's carriage-body byte encoding and octet-exact foreign
 * recovery are graded PER CLASS against the independent per-class corpus
 * vectors/carriage/&lt;class&gt;/cases.json (the round-trip identity is the authority, R-14.7). Every
 * check is fail-closed (§15): a failing object is rejected whole, returns its named error, and causes
 * no state change.
 */
public final class Carriage {
    /** Carriage classes (design.md §13.2). */
    public static final long CLASS_JSONRPC = 0;
    public static final long CLASS_HTTP = 1;
    public static final long CLASS_MSG = 2;
    public static final long CLASS_STREAM = 3;
    public static final long CLASS_DOC = 4;
    public static final long CLASS_OPAQUE = 5;

    private static final String[] CLASS_NAMES = {"JSONRPC", "HTTP", "MSG", "STREAM", "DOC", "OPAQUE"};

    private Carriage() {}

    /** The name of a class code (0..5), or "unknown". */
    public static String className(long c) {
        if (c >= 0 && c < CLASS_NAMES.length) {
            return CLASS_NAMES[(int) c];
        }
        return "unknown";
    }

    /** The body of a carriage object (design.md §13.2): {1:protocol_id,2:class,3:content_type,
     * 4:correlation,5:method,6:foreign}, where {@code foreign} is carried octet-for-octet (R-14.4). */
    public static final class CarriageBody {
        public final long protocolId;
        public final long clazz;
        public final long contentType;
        public final byte[] correlation;
        public final String method;
        public final byte[] foreign; // the foreign message, carried octet-for-octet (R-14.4)

        public CarriageBody(long protocolId, long clazz, long contentType, byte[] correlation, String method, byte[] foreign) {
            this.protocolId = protocolId;
            this.clazz = clazz;
            this.contentType = contentType;
            this.correlation = correlation.clone();
            this.method = method;
            this.foreign = foreign.clone();
        }

        /** The carriage body as a CBOR map {1..6}. */
        public Cbor.Value toValue() {
            return new Cbor.M(List.of(
                    new Cbor.Pair(new Cbor.U(1), new Cbor.U(protocolId)),
                    new Cbor.Pair(new Cbor.U(2), new Cbor.U(clazz)),
                    new Cbor.Pair(new Cbor.U(3), new Cbor.U(contentType)),
                    new Cbor.Pair(new Cbor.U(4), new Cbor.B(correlation)),
                    new Cbor.Pair(new Cbor.U(5), new Cbor.T(method)),
                    new Cbor.Pair(new Cbor.U(6), new Cbor.B(foreign))));
        }

        /** The deterministic-CBOR encoding of the carriage body. */
        public byte[] bytes() {
            return Cbor.encode(toValue());
        }
    }

    /** Reject a class code outside the defined set with a typed mapping error, never a silent drop
     * (R-14.8). Returns normally for a defined class. */
    public static void validateClass(long clazz) {
        if (clazz > CLASS_OPAQUE) {
            throw new NaalpException("MappingError", "an N-AALP semantic cannot be represented by this carriage class");
        }
    }

    /** Wrap a foreign message octet-for-octet in a carriage body (R-14.1, R-14.4). It does NOT parse,
     * canonicalize, or rewrite the foreign bytes. An undefined protocol carries under OPAQUE with an
     * experimental protocol id and zero new specification (R-18.6). An unrepresentable class is a
     * typed MappingError. */
    public static CarriageBody carry(long protocolId, long clazz, long contentType, byte[] correlation, String method, byte[] foreign) {
        validateClass(clazz);
        return new CarriageBody(protocolId, clazz, contentType, correlation, method, foreign);
    }

    /** Parse a carriage body from a CBOR value, recovering the foreign field octet-for-octet. A
     * structurally invalid body is Malformed; an unknown class is a MappingError. Fail-closed. */
    public static CarriageBody carriageFromValue(Cbor.Value v) {
        if (!(v instanceof Cbor.M m)) {
            throw new NaalpException("Malformed", "malformed carriage body");
        }
        long protocolId = 0;
        long clazz = 0;
        long contentType = 0;
        byte[] correlation = new byte[0];
        String method = "";
        byte[] foreign = null;
        for (Cbor.Pair p : m.pairs) {
            if (!(p.k instanceof Cbor.U k)) {
                throw new NaalpException("Malformed", "non-uint carriage key");
            }
            switch ((int) k.v) {
                case 1:
                    protocolId = requireUint(p.val);
                    break;
                case 2:
                    clazz = requireUint(p.val);
                    break;
                case 3:
                    contentType = requireUint(p.val);
                    break;
                case 4:
                    correlation = requireBstr(p.val);
                    break;
                case 5:
                    if (!(p.val instanceof Cbor.T t)) {
                        throw new NaalpException("Malformed", "method is not a text string");
                    }
                    method = t.v;
                    break;
                case 6:
                    foreign = requireBstr(p.val);
                    break;
                default:
                    throw new NaalpException("Malformed", "unexpected carriage key " + k.v);
            }
        }
        if (foreign == null) {
            throw new NaalpException("Malformed", "carriage body has no foreign field");
        }
        validateClass(clazz);
        return new CarriageBody(protocolId, clazz, contentType, correlation, method, foreign);
    }

    private static long requireUint(Cbor.Value v) {
        if (!(v instanceof Cbor.U u)) {
            throw new NaalpException("Malformed", "expected a uint");
        }
        return u.v;
    }

    private static byte[] requireBstr(Cbor.Value v) {
        if (!(v instanceof Cbor.B b)) {
            throw new NaalpException("Malformed", "expected a byte string");
        }
        return b.v;
    }

    /** Protocol id ranges (design.md §13.4): reserved 0x00, standards 0x01-0x0F,
     * experimental 0x10-0x7F (no registration), private 0x80-0xFF; a value above one octet
     * is invalid. */
    public static String protocolRange(long id) {
        if (id == 0x00) {
            return "reserved";
        }
        if (id <= 0x0F) {
            return "standards";
        }
        if (id <= 0x7F) {
            return "experimental";
        }
        if (id <= 0xFF) {
            return "private";
        }
        return "invalid"; // protocol_id is one octet
    }

    /** Records whether a carried message was actually delivered. A report never claims delivery it did
     * not achieve (R-14.8). */
    public static final class DeliveryReport {
        public final boolean delivered;

        public DeliveryReport(boolean delivered) {
            this.delivered = delivered;
        }
    }

    /** Produce a delivery report from the below-foreign outcome: a failed delivery yields
     * delivered=false and NotDelivered — never a false "delivered" (R-14.8). */
    public static DeliveryReport report(boolean deliveredBelow) {
        if (!deliveredBelow) {
            throw new NaalpException("NotDelivered", "a below-foreign failure; the message was not delivered");
        }
        return new DeliveryReport(true);
    }

    /** The authorizing principal of a carriage object: the N-AALP signer of the object (envelope field
     * 5), never any foreign principal named inside the foreign bytes (R-14.6). It reads only the
     * signed envelope, never the carried foreign message. */
    public static byte[] carriageAuthority(Envelope.Object o) {
        return o.signer;
    }
}

File: impl/kotlin/src/main/kotlin/sh/bubblefish/naalp/Carriage.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

/**
 * N-AALP C12 foreign carriage by class for the Kotlin SDK (design.md section 13; R-14.1..14.8,
 * R-18.6).
 *
 * N-AALP carries a foreign agent protocol by wrapping its message, octet-for-octet, in a signed
 * N-AALP carriage object whose effect, safety, identity, and audit apply, and whose foreign body is
 * interpreted by a carriage CLASS -- not a bespoke per-protocol mapping (R-14.1). There are five
 * structured classes (JSONRPC, HTTP, MSG, STREAM, DOC) plus a universal OPAQUE class that makes any
 * protocol -- including one nobody has defined -- carriable immediately on an experimental protocol id
 * with no registration (R-14.1, R-18.6). The foreign field is carried VERBATIM and MUST NOT be
 * re-serialized, canonicalized, summarized, or rewritten (R-14.4); N-AALP metadata is carried around
 * it, never inside it. The carriage object's signer remains the authority -- a foreign identity never
 * becomes an N-AALP authorization identity (R-14.6).
 *
 * An independent transcription of impl/go/carriage (a second reference is
 * impl/python/naalp/carriage.py); each carriage class is graded against its own per-class oracle at
 * vectors/carriage/<class>/cases.json. Carriage bodies are UNSIGNED (the six per-class body_hex ARE
 * the byte parity); the R-14.6 identity-containment property is demonstrated over a real signed
 * Envelope object. Every check is fail-closed.
 */
object Carriage {
    // Carriage classes (design.md section 13.2).
    const val CLASS_JSONRPC = 0L
    const val CLASS_HTTP = 1L
    const val CLASS_MSG = 2L
    const val CLASS_STREAM = 3L
    const val CLASS_DOC = 4L
    const val CLASS_OPAQUE = 5L

    private val CLASS_NAMES = arrayOf("JSONRPC", "HTTP", "MSG", "STREAM", "DOC", "OPAQUE")

    /** The name of a class code (0..5), or "unknown". */
    fun className(c: Long): String = if (c in 0 until CLASS_NAMES.size.toLong()) CLASS_NAMES[c.toInt()] else "unknown"

    /**
     * The body of a carriage object (design.md section 13.2). [klass] is the carriage class (spelled
     * `klass` because `class` is a Kotlin keyword); [foreign] is the foreign message carried
     * octet-for-octet (R-14.4).
     */
    class CarriageBody(
        val protocolId: Long,
        val klass: Long,
        val contentType: Long,
        correlation: ByteArray,
        val method: String,
        foreign: ByteArray,
    ) {
        val correlation: ByteArray = correlation.copyOf()
        val foreign: ByteArray = foreign.copyOf()

        /** The carriage body as a CBOR map {1: protocol_id, 2: class, 3: content_type,
         *  4: correlation, 5: method, 6: foreign}. */
        fun toValue(): Cbor.Value = Cbor.M(
            listOf(
                Cbor.Pair(Cbor.U(1), Cbor.U(protocolId)),
                Cbor.Pair(Cbor.U(2), Cbor.U(klass)),
                Cbor.Pair(Cbor.U(3), Cbor.U(contentType)),
                Cbor.Pair(Cbor.U(4), Cbor.B(correlation)),
                Cbor.Pair(Cbor.U(5), Cbor.T(method)),
                Cbor.Pair(Cbor.U(6), Cbor.B(foreign))
            )
        )

        /** The deterministic-CBOR encoding of the carriage body. */
        fun bytes(): ByteArray = Cbor.encode(toValue())
    }

    /** Reject a class code outside the defined set with a typed mapping error, never a silent drop
     *  (R-14.8). Returns on a representable class. */
    fun validateClass(klass: Long) {
        if (klass > CLASS_OPAQUE || klass < 0) {
            throw NaalpException("MappingError", "an N-AALP semantic cannot be represented by this carriage class")
        }
    }

    /**
     * Wrap a foreign message octet-for-octet in a carriage body (R-14.1, R-14.4). It does not parse,
     * canonicalize, or rewrite the foreign bytes. An undefined protocol carries under OPAQUE with an
     * experimental protocol id and zero new specification (R-18.6).
     */
    fun carry(protocolId: Long, klass: Long, contentType: Long, correlation: ByteArray, method: String, foreign: ByteArray): CarriageBody {
        validateClass(klass)
        return CarriageBody(protocolId, klass, contentType, correlation, method, foreign)
    }

    /**
     * Parse a carriage body from a CBOR value, recovering the foreign field octet-for-octet. A
     * structurally invalid body is Malformed; an unknown class is a MappingError. Only the mandatory
     * foreign field (key 6) is required; fields 1-5 default to their zero values (matching Go/Python).
     * An unknown field key is rejected Malformed.
     */
    fun carriageFromValue(v: Cbor.Value): CarriageBody {
        if (v !is Cbor.M) throw NaalpException("Malformed", "carriage body is not a map")
        var protocolId = 0L
        var klass = 0L
        var contentType = 0L
        var correlation = ByteArray(0)
        var method = ""
        var foreign = ByteArray(0)
        var haveForeign = false
        for (p in v.pairs) {
            val k = p.k
            if (k !is Cbor.U) throw NaalpException("Malformed", "non-uint carriage key")
            when (k.v) {
                1L -> {
                    val u = p.v as? Cbor.U ?: throw NaalpException("Malformed", "protocol_id not a uint")
                    protocolId = u.v
                }
                2L -> {
                    val u = p.v as? Cbor.U ?: throw NaalpException("Malformed", "class not a uint")
                    klass = u.v
                }
                3L -> {
                    val u = p.v as? Cbor.U ?: throw NaalpException("Malformed", "content_type not a uint")
                    contentType = u.v
                }
                4L -> {
                    val b = p.v as? Cbor.B ?: throw NaalpException("Malformed", "correlation not a bstr")
                    correlation = b.v
                }
                5L -> {
                    val t = p.v as? Cbor.T ?: throw NaalpException("Malformed", "method not a tstr")
                    method = t.v
                }
                6L -> {
                    val b = p.v as? Cbor.B ?: throw NaalpException("Malformed", "foreign not a bstr")
                    foreign = b.v
                    haveForeign = true
                }
                else -> throw NaalpException("Malformed", "unknown carriage field ${k.v}")
            }
        }
        if (!haveForeign) throw NaalpException("Malformed", "carriage body missing the mandatory foreign field")
        validateClass(klass)
        return CarriageBody(protocolId, klass, contentType, correlation, method, foreign)
    }

    /**
     * The protocol-id range (design.md section 13.4): reserved 0x00, standards 0x01-0x0F,
     * experimental 0x10-0x7F (no registration), private 0x80-0xFF.
     * protocol_id is one octet; anything wider is invalid.
     */
    fun protocolRange(id: Long): String = when {
        id == 0x00L -> "reserved"
        id <= 0x0FL -> "standards"
        id <= 0x7FL -> "experimental"
        id <= 0xFFL -> "private"
        else -> "invalid"
    }

    /** Records whether a carried message was actually delivered. A report never claims delivery it did
     *  not achieve (R-14.8). */
    class DeliveryReport(val delivered: Boolean)

    /**
     * Produce a delivery report from the below-foreign outcome: a failed delivery throws NotDelivered
     * (never a false "delivered"); a success returns a delivered report (R-14.8).
     */
    fun report(deliveredBelow: Boolean): DeliveryReport {
        if (!deliveredBelow) throw NaalpException("NotDelivered", "a below-foreign failure; the message was not delivered")
        return DeliveryReport(true)
    }

    /**
     * The authorizing principal of a carriage object: the N-AALP signer of the object (envelope field
     * 5), never any foreign principal named inside the foreign bytes (R-14.6). It reads only the
     * signed envelope, never the carried foreign message.
     */
    fun carriageAuthority(obj: Envelope.Object): ByteArray = obj.signer
}
